torch/distributed/_tools/mem_tracker.py

```diff
@@ -0,0 +1,920 @@
+        """
+        Updates and returns the memory consumed if the storage size has changed.
+
+        Args:
+        st (torch.UntypedStorage): The tensor storage to check for size updates.
+
+        Returns:
+        int: The updated memory consumed in bytes.
+        """
+        if st.size() != self.size:
+            self.size = st.size()
+            self.mem_consumed = self._calculate_mem_consumed()
+        return self.mem_consumed
+
+    @staticmethod
+    def get_untyped_storages(t: torch.Tensor) -> Set[torch.UntypedStorage]:
+        """
+        Recursively extracts untyped storages from a tensor or its subclasses.
+
+        Args:
+        t (torch.Tensor): The tensor to extract storages from.
+
+        Returns:
+        Set[torch.UntypedStorage]: A set of untyped storages.
+        """
+        unflattend_tensors = [t]
+        flattened_tensor_storages = set()
+        while len(unflattend_tensors) > 0:
+            obj = unflattend_tensors.pop()
+            if is_traceable_wrapper_subclass(obj):
+                attrs, _ = obj.__tensor_flatten__()  # type: ignore[attr-defined]
+                unflattend_tensors.extend([getattr(obj, attr) for attr in attrs])
+            else:
+                flattened_tensor_storages.add(obj.untyped_storage())
+        return flattened_tensor_storages
+
+    @classmethod
+    def create_winfo(
+        cls,
+        st: torch.UntypedStorage,
+        device: torch.device,
+        reftype: _RefType,
+        callback: Optional[Callable[[Self, weakref.ref], Any]] = None,
+    ) -> Tuple[Self, weakref.ref]:
+        """
+        Creates a new ``_WeakRefInfo`` instance and a weak reference to a ``torch.UntypedStorage`` object,
+        optionally attaching a callback to the weak reference.
+
+        Args:
+            st (torch.UntypedStorage): The storage object for which to create the weak reference info.
+            device (torch.device): The device associated with the storage object.
+            reftype (_RefType): The type of reference, used to categorize the storage.
+            callback (Optional[Callable[[Self, weakref.ref]]]): A callback function that is called when
+                the storage object is about to be finalized (garbage collected). The callback function
+                should accept two arguments: the ``_WeakRefInfo`` instance and the weak reference to the storage.
+        Returns:
+            Tuple[Self, weakref.ref]: A tuple containing the newly created ``_WeakRefInfo`` instance and the
+            weak reference to the storage object. The weak reference may have an attached callback if provided.
+        """
+
+        winfo = cls(st.size(), st.element_size(), device, reftype)
+        w_st = weakref.ref(st, partial(callback, winfo) if callback else None)
+        return winfo, w_st
+
+
+def _get_mem_divisor(units: str) -> int:
+    return {"B": 1, "KB": 2**10, "MB": 2**20, "GB": 2**30}.get(units, 1)
```

Comment:
nit: should these be `KiB`, `MiB`, and `GiB` if we are using base 2?

Reply:
Yeah, will fix it